### SNP, indel, and SV detection using whole haplotype genome alignment

To call heterozygous sites between the two haploid sequences, independent of the GenomeScope calculation, we first performed a Mummer alignment. Because our assemblies span most repetitive sequences, repeat-masking treatment was not necessary before conducting Mummer alignment. 


### split genome by chrs

two haplotype assemblies will be splited into different folder by chromosome ID.

```shell
python3 split_genome_bychrs.py mCalJac1.maternal.chrs.fa mCalJac1.paternal.chrs.fa
```
Note: the fasta sequence header is formatted like 'mat_Chr1' for maternal, and 'patChr1' for maternal.

### Mummer alignment, and SNV, small indel detection

```shell
cat chr.list|while read a
do
	cd $a
	software/MUMmer-3.23/nucmer -maxmatch -l 100 -c 500 -p $a mat_$a.fa pat_$a.fa
	software/MUMmer-3.23/dnadiff -d $a.delta -p $a.diff

	# $a.diff.snps contains the SNVs and INDELs
	mv $a.diff.snps $a.diff.var

	# split variations into two files.
	python3 ../split_snp_indel.py $a.diff.var $a.diff

	# change indel format for easy reading
	python3 indel_statistic.py $a.diff.indel $a

	# split small and longer indels
	awk '$6<=50' $a.diff.indel.report.txt >$a.mummer.sm.indel
	awk '$6>50' $a.diff.indel.report.txt >$a.mummer.lg.indel

	cd ..
done
```

# Indel

## how to detect indel between two alleles
indel result can be found in mummer directory, which generated by nucmer program:


have a look at indel result:

~~~~~text
7270	T	.	1939	3	1939	218136636	216975769	1	1	mat_Chr1	pat_Chr1
8102	T	.	2770	1	2770	218136636	216975769	1	1	mat_Chr1	pat_Chr1
8103	G	.	2770	1	2770	218136636	216975769	1	1	mat_Chr1	pat_Chr1
~~~~~

it's one site in a line, not well to read, so convert it to one record in a line:

~~~~~~~~~~bash
python3 bin/indel_statistic_RefPos.py $a.diff.indel $a
~~~~~~~~~~

now look again:

~~~~~text
# S. means ref is string (> 1bp) and a deletion in query
#Chrom	R	type	subType	length	R_base	Q_base
Chr1.diff	7270	insert	T.	1	T	.
Chr1.diff	8102	delete	S.	2	TG	.
~~~~~

## filter indel base 10x read mapping
A large number of Indels are adjacent with repeative sequences. especially polynucleotides which
indicates that many Indels are probably false. So the validation using shotgun reads is very important.

can use below scripts to filter based on depth

```
# small indel supported by 10x at least 3 reads
	# here $a.10x.base.xls is calculated from bam file (10x reads mapping to maternal assembly)
	python3 ../filter_indel_by_depth.py $a.mummer.sm.indel xxx/GATK_10X/$a/$a.10x.base.xls  $a.mummer.sm.10xsupport.indel 1>sm.indel.check.log

	# large indel supported by Pacbio long reads at least 3 reads
	# here $a.pat2mat.cor.base.xls is calculated form bam file (paternal pacbio reads mapping to maternal assembly)
	echo "python3 ../filter_indel_by_depth.py $a.mummer.lg.indel xxx/pat2mat/$a/$a.pat2mat.cor.base.xls  $a.mummer.lg.Pbsupport.indel 1>lg.indel.check.log " > $a.check.lg.sh
	# run $a.check.lg.sh

```


## small indel (< 50 bp) and longer indel (>= 50 bp)

Mummer also report some indels that are longer than 50 bp, this dataset will combine with INDELs detected by [svmu](../Structure_variations/README.md.html)
